clients/spark-extensions-base/src/main/scala/org/apache/spark/sql/execution/datasources/v2/NessieUtils.scala

```diff
@@ -194,18 +194,37 @@ object NessieUtils {
       .build(classOf[NessieApiV1])
   }
 
+  /** @param currentCatalog
+    *   The current Spark catalog
+    * @param catalog
+    *   The catalog to configure for Spark
+    * @param ref
+    *   The reference to configure for Spark
+    * @param configureRefAtHash
+    *   Whether to configure the ref at its given hash. Note that this should
+    *   only be done when reading data from the given ref at its particular
+    *   hash.
+    */
   def setCurrentRefForSpark(
       currentCatalog: CatalogPlugin,
       catalog: Option[String],
-      ref: Reference
+      ref: Reference,
+      configureRefAtHash: Boolean
   ): Unit = {
     val catalogName = catalog.getOrElse(currentCatalog.name)
     val catalogImpl =
       SparkSession.active.sessionState.catalogManager.catalog(catalogName)
     SparkSession.active.sparkContext.conf
       .set(s"spark.sql.catalog.$catalogName.ref", ref.getName)
-    SparkSession.active.sparkContext.conf
-      .set(s"spark.sql.catalog.$catalogName.ref.hash", ref.getHash)
+    if (configureRefAtHash) {
+      // we only configure ref.hash if we're reading data
+      SparkSession.active.sparkContext.conf
+        .set(s"spark.sql.catalog.$catalogName.ref.hash", ref.getHash)
+    } else {
+      // we need to clear it in case it was previously set
+      SparkSession.active.sparkContext.conf
```

Comment:
I am not sure whether we need to clear this conf or set it to the current head of this reference.
I think it depends on catalog implementation.

Hope you have verified with end to end changes. 

These order of operations should work:
```
use reference@hash/timestamp ; --(points to old commit)
select * from table; -- data from old commit
insert to the table; --?? may be should not work.
use reference; -- (points to head)
select * from table; -- data from head
insert to the table; --should work.
select * from table@hash/timestamp ; -- data from old commit
```

Reply:
Resetting to head will become same as master branch code. So, I guess. This PR should help us to handle things at Iceberg side. 
LGTM.

Reply:
yes resetting is actually required (I verified with end-to-end testing)